src/string.rs

```diff
@@ -1 +1,289 @@
+            }
+        }
+        buf.copy_from_slice(&sum_l);
+        for i in (0..n).rev() {
+            let v = sa[i];
+            if v >= 2 && ls[v - 2] {
+                buf[s[v - 2] + 1] -= 1;
+                sa[buf[s[v - 2] + 1]] = v - 1;
+            }
+        }
+    };
+    // origin: 1
+    let mut lms_map = vec![0; n + 1];
+    let mut m = 0;
+    for i in 1..n {
+        if !ls[i - 1] && ls[i] {
+            lms_map[i] = m + 1;
+            m += 1;
+        }
+    }
+    let mut lms = Vec::with_capacity(m);
+    for i in 1..n {
+        if !ls[i - 1] && ls[i] {
+            lms.push(i);
+        }
+    }
+    assert_eq!(lms.len(), m);
+    induce(&mut sa, &lms);
+
+    if m > 0 {
+        let mut sorted_lms = Vec::with_capacity(m);
+        for &v in &sa {
+            if lms_map[v - 1] != 0 {
+                sorted_lms.push(v - 1);
+            }
+        }
+        let mut rec_s = vec![0; m];
+        let mut rec_upper = 0;
+        rec_s[lms_map[sorted_lms[0]] - 1] = 0;
+        for i in 1..m {
+            let mut l = sorted_lms[i - 1];
+            let mut r = sorted_lms[i];
+            let end_l = if lms_map[l] < m { lms[lms_map[l]] } else { n };
+            let end_r = if lms_map[r] < m { lms[lms_map[r]] } else { n };
+            let mut same = true;
+            if end_l - l != end_r - r {
+                same = false;
+            } else {
+                while l < end_l {
+                    if s[l] != s[r] {
+                        break;
+                    }
+                    l += 1;
+                    r += 1;
+                }
+                if l == n || s[l] != s[r] {
+                    same = false;
+                }
+            }
+            if !same {
+                rec_upper += 1;
+            }
+            rec_s[lms_map[sorted_lms[i]] - 1] = rec_upper;
+        }
+
+        let rec_sa = sa_is::<T>(&rec_s, rec_upper);
+        for i in 0..m {
+            sorted_lms[i] = lms[rec_sa[i]];
+        }
+        induce(&mut sa, &mut sorted_lms);
+    }
+    for i in 0..n {
+        sa[i] -= 1;
+    }
+    sa
+}
+
+fn sa_is_i32<T: Threshold>(s: &[i32], upper: i32) -> Vec<usize> {
+    let s: Vec<usize> = s.iter().map(|&x| x as usize).collect();
+    sa_is::<T>(&s, upper as usize)
+}
+
+pub fn suffix_array_manual(s: &[i32], upper: i32) -> Vec<usize> {
+    assert!(upper >= 0);
+    for &elem in s {
+        assert!(0 <= elem && elem <= upper);
+    }
+    sa_is_i32::<DefaultThreshold>(s, upper)
+}
+
+pub fn suffix_array_arbitrary<T: Ord>(s: &[T]) -> Vec<usize> {
+    let n = s.len();
+    let mut idx: Vec<usize> = (0..n).collect();
+    idx.sort_by_key(|&i| &s[i]);
+    let mut s2 = vec![0; n];
+    let mut now = 0;
+    for i in 0..n {
+        if i > 0 && s[idx[i - 1]] != s[idx[i]] {
+            now += 1;
+        }
+        s2[idx[i]] = now;
+    }
+    sa_is_i32::<DefaultThreshold>(&s2, now)
+}
+
+pub fn suffix_array(s: impl IntoIterator<Item = char>) -> Vec<usize> {
```

Comment:
Maybe include `ascii` in the naming (plus use 127 as the upper bound) or use `u8`?

Reply:
I'm thinking about changing this function's signature to something like `fn suffix_array(s: &str)` because this function can't take `String` or `&str` directly.